### PR TITLE
Improve perceived performance by preloading the template parts defined in Theme JSON when in Site Editor

### PR DESCRIPTION
**Pull request by getdave**
## What?

Improves the perceived loading performance of the Site Editor by (selectively) _preloading_  template parts (see `How` for details).

## Why?

Most of the perceived sluggish-ness of the Site Editor loading is due to having to wait for template parts to resolve from the network.

For example on TT3, when the editor loads requests are dispatched for 

- `/wp/v2/template-parts/twentytwentythree//header?context=edit`
- `/wp/v2/template-parts/twentytwentythree//footer?context=edit`

The former is particularly jarring as the whole header area becomes a loading spinner until the network resolves. This is very noticeable on slower connections (take a minute to test it).

This is further compounded by the fact that most `header` template parts will contain a Navigation block which has it's own loading state which is dependent on loading entities over the network. As the Nav block will not be rendered in the editor until the template part resolves, this means we get a loading waterfall effect where we see

- template part loading
- template part resolves
- navigation block starts loading
- navigation block resolves

 This is not a great experience.

## How?

By preloading requests for the templates parts defined by the `theme.json` we can ensure these parts are available _immediately_ in the `@wordpress/api-fetch` network cache upon editor load. 

This means the editor never dispatches the network `fetch` request and instead [core data] _immediately_ returns the entity data  for the template part directly. 

This results in almost **no loading spinners** in the header. When combined with https://github.com/WordPress/gutenberg/pull/48683 this should have a major impact on improve the perceived performance.

### Criteria for Preloading a Template

When determining which template parts will be preloaded the following criteria apply:

- must be defined in `theme.json` under the `templateParts` key.
- must include a `area` property set to `header` - these are templates being specifically denoted as being designed for the "Header" Template Area of a site.
- within the first 3 templates - we hard limit to 3 preloaded templates by default to avoid unwanted edge cases. This can be filtered using the `gutenberg_template_part_preloading_limit` filter. 

## ⚠️ Potential Considerations

- may _slightly_ increase TTFB in the Site Editor (only) - looking at Gutenberg Performance tests (see CI for this PR) changes seem negligible ([example](https://github.com/WordPress/gutenberg/pull/48924/files#r1138941821))
- ~may result in 404 requests on the server if the `header` or `footer` parts do not exist.~ concern no longer applies

## Testing Instructions

- Activate TT3
- Throttle your connection a bit.
- Open Site Editor
- See no loading spinners for the site Header.
- Compare with `trunk` where there are loading spinners which take a long time to resolve.

#### Before

Notice the loading spinner for the template part.

https://user-images.githubusercontent.com/444434/223717310-713da1bd-f22e-4599-9081-e1d8a06345ff.mp4

#### After

Notice no spinner for the template part.

https://user-images.githubusercontent.com/444434/223717115-dccc7ab0-c7d1-4b1b-84b2-5f4c747d9302.mp4

### CODE REVIEW

**Comment by getdave** [comment]
Related https://github.com/WordPress/gutenberg/pull/48683

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * @return array
+ */
+function gutenberg_preload_template_parts( $preload_paths, $context ) {
+
+	// Limit to the Site Editor.
+	if ( ! empty( $context->name ) && 'core/edit-site' !== $context->name ) {
+		return $preload_paths;
+	}
+
+	$theme_slug = wp_get_theme()->get_stylesheet();
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.3/template-parts-preloading.php]:
```suggestion
	$theme_slug = get_stylesheet();
```

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * Patches resources loaded by the block editor page
+ * to include Template Part posts.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Preloads requests needed for common Template Part posts
+ *
+ * @param array $preload_paths    Preload paths to be filtered.
```
Comment (tyxla) [inline lib/compat/wordpress-6.3/template-parts-preloading.php:12]: We're missing a `@param` for the `$context` argument

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		return $preload_paths;
+	}
+
+	$theme_slug = wp_get_theme()->get_stylesheet();
+
+	$template_parts_rest_route = rest_get_route_for_post_type_items(
+		'wp_template_part'
+	);
+
+	$standard_template_parts = array( 'header', 'footer' );
```
Comment (tyxla) [inline lib/compat/wordpress-6.3/template-parts-preloading.php]: This asserts that the template part names will always be the same. Given that those can often change and are dynamic, I think this optimization won't help us much. 

Furthermore, this asserts that these are the most widely used template parts, which often won't be the case. I've seen some themes with multiple headers and footers (usually one for the home page, one for the inner pages), and this may not only catch any of them; but also incorrectly assume that these are the most important ones. In the scenario of "one header/footer pair for home page, one pair for all other pages", maybe those represent the ones used for the home page and will end up being preloaded regardless that the other header/footer pair is used more often. 

I wonder if we could try to be smarter about those and preload template parts that are actually widely used across the current templates.

**Review by tyxla — COMMENTED** [review]
Thank you for the great work @getdave! 

I've left some questions.

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 3 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php]: > This asserts that the template part names will always be the same. Given that those can often change and are dynamic, I think this optimization won't help us much.

Yes this is entirely true. My rationale is that the majority of Themes will likely use `header` and `footer` for their Template Part names for those locations. TBH we could ditch footer and just load `header.

> I've seen some themes...

I think this is the key statement. Optimise for the majority but don't break it for the minority. Will this make the experience worse for those edge case themes and does that outweigh the benefit for the majority?

> I wonder if we could try to be smarter about those and preload template parts that are actually widely used across the current templates.

That would require another request for templates. It starts to become self defeating.

**Comment by getdave** [comment]
Thanks for the review @tyxla. I take your points and they were my main concerns when raising this PR (see description).

It comes down to the same as the other PR for Navigation menus. Does the benefit for the majority outweight any downside for the minority.

My guess is that this PR will end up being closed as it's too un-targetted to be palatable.

**Comment by tyxla** [comment]
> Thanks for the review @tyxla. I take your points and they were my main concerns when raising this PR (see description).
> 
> It comes down to the same as the other PR for Navigation menus. Does the benefit for the majority outweight any downside for the minority.

I fully agree with that, and from my perspective, I can't really tell, because my experience with template parts and themes isn't extensive enough. I just wanted to bring up what I know and would prefer to leave the final decision to someone who can better weigh the tradeoffs you outlined.

**Comment by getdave** [comment]
Thanks @tyxla. I appreciate you providing that perspective. It's important.

I'm going to try one more thing to see if we can improve this implementation.

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 4]
```diff
@@ -19,17 +19,26 @@ function gutenberg_preload_template_parts( $preload_paths, $context ) {
 		return $preload_paths;
 	}
 
-	$theme_slug = wp_get_theme()->get_stylesheet();
+	// Get any template parts defined in theme.json.
+	$theme_json_template_parts = WP_Theme_JSON_Resolver::get_merged_data()->get_template_parts();
```
Comment (getdave) [inline lib/compat/wordpress-6.3/template-parts-preloading.php:26]: @tyxla What do you think about this more selective approach? Here we only preload those template parts that are defined **by the Theme** in the Theme JSON.

Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: Example

https://github.com/WordPress/twentytwentythree/blob/11440a5b38bae6882fc2c00b703d4bb95420a0f5/theme.json#L712-L733

```
	"templateParts": [
		{
			"area": "header",
			"name": "header",
			"title": "Header"
		},
		{
			"area": "footer",
			"name": "footer",
			"title": "Footer"
		},
		{
			"area": "uncategorized",
			"name": "comments",
			"title": "Comments Template Part"
		},
		{
			"area": "uncategorized",
			"name": "post-meta",
			"title": "Post Meta"
		}
	]
```

Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: @scruffian I don't really understand the value of having a `templateParts` definition in Theme JSON but Themes are doing it. It could act as a optimising with the Theme "telling" the editor how to optimise itself.

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 1 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php]: ✅

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 4 above)
Comment (tyxla) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: This looks like a sensible approach 👍 

I assume that `WP_Theme_JSON_Resolver::get_merged_data()` caches data in runtime so it doesn't have to be processed on every call? 

Have you been able to measure its performance impact? It seems to be that it would load pretty quickly because the theme.json data has already been parsed and is available in the memory, but I wanted to double-check.

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+	$theme_json_template_part_slugs = array_keys( $theme_json_template_parts );
+
+	$theme_slug = get_stylesheet();
+
+	$template_parts_rest_route = rest_get_route_for_post_type_items(
+		'wp_template_part'
+	);
+
+	$standard_template_parts = array( 'header', 'footer' );
```
Comment (tyxla) [inline lib/compat/wordpress-6.3/template-parts-preloading.php]: We should be able to remove this variable now.

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+	foreach ( $theme_json_template_part_slugs as $template_part_slug ) {
+		$preload_paths[] = array(
+			$template_parts_rest_route . "/$theme_slug//$template_part_slug?context=edit",
```
Comment (tyxla) [inline lib/compat/wordpress-6.3/template-parts-preloading.php]: I'd strongly recommend against parsing variables in strings, since it could be similar to what `eval()` does, and it's not recommended to be used. We could do this:

```suggestion
			$template_parts_rest_route . '/' . $theme_slug . '//' . $template_part_slug . '?context=edit',
```

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 4 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: I'm exploring both the built in Gutenberg perf tests `npm run test:performance`

...and also

https://github.com/GoogleChromeLabs/wpp-research/tree/main 

...on the advice of @adamsilverstein.

Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: Performance test results from CI running on this PR (having rebased against `trunk`):

```
>> site-editor

┌──────────────────────┬──────────────────────────────────────────┬──────────────┐
│       (index)        │ f060a497fa86698a5f7e9bc92d168a8a71225fad │    trunk     │
├──────────────────────┼──────────────────────────────────────────┼──────────────┤
│    serverResponse    │                '138.1 ms'                │  '140.7 ms'  │
│      firstPaint      │                '482.2 ms'                │  '400.7 ms'  │
│   domContentLoaded   │                '789.2 ms'                │  '689.8 ms'  │
│        loaded        │                '790.5 ms'                │ '690.93 ms'  │
│ firstContentfulPaint │               '879.53 ms'                │  '782.9 ms'  │
│      firstBlock      │               '1972.53 ms'               │ '1872.07 ms' │
│         type         │                '74.3 ms'                 │  '70.8 ms'   │
│       minType        │                '62.94 ms'                │  '60.88 ms'  │
│       maxType        │               '496.34 ms'                │ '148.81 ms'  │
└──────────────────────┴──────────────────────────────────────────┴──────────────┘
```

`serverResponse ` is the most important I believe. 

Can also be run locally with `npm run test:performance`.

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 3 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php]: This has been fixed by only loading template parts defined in `theme.json`.

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 5 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php]: ✅

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 4 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: I've [reached out in the `#core-performance` channel on WordPress Slack](https://wordpress.slack.com/archives/C02KGN5K076/p1679066571642589) for some support here. I know @felixarntz was involved in a previous discussion about preloading media so perhaps he has a view?

**Comment by getdave** [comment]
Looping in @youknowriad here as lead architect in case he has a strong opinion on whether preloading only the templateParts defined in the Theme JSON would be a fair compromise to help improve the perceived loading performance of the Site Editor.

File: lib/compat/wordpress-6.3/template-parts-preloading.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ */
+function gutenberg_preload_template_parts( $preload_paths, $context ) {
+
+	// Limit to the Site Editor.
+	if ( ! empty( $context->name ) && 'core/edit-site' !== $context->name ) {
+		return $preload_paths;
+	}
+
+	// Get any template parts defined in theme.json.
+	$theme_json_template_parts = WP_Theme_JSON_Resolver::get_merged_data()->get_template_parts();
```
Comment (getdave) [inline lib/compat/wordpress-6.3/template-parts-preloading.php:26]: @anton-vlasenko @ironprogrammer Do you have any idea of how I might be able to mock `WP_Theme_JSON_Resolver` in a test? I want to write tests but not sure how I could make that testable.

**Comment by youknowriad** [comment]
> Looping in @youknowriad here as lead architect in case he has a strong opinion on whether preloading only the templateParts defined in the Theme JSON would be a fair compromise to help improve the perceived loading performance of the Site Editor.

Yeah, I think that's a good starting point but I'm curious about themes that define multiple headers and multiple footers... Is there a way to get only one header, one footer and one navigation menu (the main ones)

**Comment by getdave** [comment]
> Is there a way to get only one header, one footer and one navigation menu (the main ones)

Some things I've thought about but not explored:

- getting all the `wp_template`s, parsing the blocks and checking which template_parts are in use.
- adding post meta to `wp_template_part` (when added to a template) to indicate it is "in use" and only preloading those

However, I'm mindful of the trade offs for these optimisations. If we're making more queries to avoid making queries for all templates then it seems counter intuative.

That said, if there is some heuristic that can allows us to determine the "main" header/footer then that would be ideal. I'm just not sure there's anything to go on.

**Comment by getdave** [comment]
@youknowriad I wonder if we should also load any template parts that either:

- contain the string `header` in the slug

AND/OR

- are in the `header` template part `area`

This would involve adding another item to the preload paths along these lines:

```
template-parts?area=header
```

Reason? The current implementation will not preload custom templates created by the user in the editor _unless_ they edited one of the Theme's template parts.

**Comment by youknowriad** [comment]
Yeah, it's not clear,  I guess if there's only a couple (or a max threshold we fix of template areas assigned to the "header" and "footer" areas, it should be fine. But I can see a world where some sites downloads a high number of headers and footers and the preloading becomes a problem. So maybe limit this to the first two headers and footers or something like that.

**Comment by getdave** [comment]
Research

- preloading only populates a cache for `apiFetch` on a 1:1 match basis of request `path` against preload cache key.
- getEntityRecords will resolve for `getEntityRecord` so that apiFetch is not dispatched for those records already in the core-data state cache.

As such preloading `template-parts?area=header` will _not_ avoid apiFetch requests for `template-parts//{{theme_slug}}//{{templarte_part_slug}}` because the core-data state has no record of having resolved the preloaded data for all template parts. Therefore the request is still dispatched.

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 7 above)
Comment (oandregal) [inline lib/compat/wordpress-6.3/template-parts-preloading.php:26]: In the plugin, we need to use the classes bundled by Gutenberg: `WP_Theme_JSON_Resolver_Gutenberg` instead of `WP_Theme_JSON_Resolver`.

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 4 above)
Comment (oandregal) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: I haven't manually tried this change, though running the performance tests locally this PR performs worse in every metric but typing :thinking: 

- this PR:

230320-branch-preloading-template-parts


- trunk

image

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 7 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: @oandregal On another note it would be nice if there was a stable global function which I could use instead of `::get_merged_data()->get_template_parts()`. Is that a good idea?

Comment (oandregal) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: Absolutely! We should not use the private classes in places other than the high-level functions (`get_global_settings`, etc.). It's tracked as part of https://github.com/WordPress/gutenberg/issues/45171 (though it's up for anyone to pick it up):

> Implement wp_get_custom_templates_from_theme: access to customTemplates defined by the theme

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 4 above)
Comment (oandregal) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: I now have tested this locally by throttling to "slow 3g" and loading http://localhost:8889/wp-admin/site-editor.php (which means loading "browse mode", not the site editor in edit mode).

In that setup, the canvas contents take ages to load. Within the canvas, the components that took the most time were navigation and query block. The header template was shown almost immediately, both in trunk and this PR.

Perhaps the stats are a bit misleading for this change, because none measure "everything (or the header) has been rendered". First Contentful Paint is probably rendering the site title "Mindblown: a blog about philosophy". I don't see that we track Largest Contentful Paint, though given the importance of the header for TT3 in the whole template, I'd expect it to be affected by this change.

TLDR: I don't see any positive effect of this change. However! If other people are, I don't want to block this. I'll leave space for others to comment.

**Comment by getdave** [comment]
> But I can see a world where some sites downloads a high number of headers and footers and the preloading becomes a problem. So maybe limit this to the first two headers and footers or something like that.

I've updated the PR to only preload the first `3` template parts that are listed in the `theme.json` under the `templateParts` key and whose `area` property is `header`.

So we

- selective in the template parts we are preloading
- avoid making too many requests
- avoid making queries for queries
- optimise for the most common use case

How do you feel about this now?

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 2 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:12]: ✅

File: lib/compat/wordpress-6.3/template-parts-preloading.php (hunk 4 above)
Comment (getdave) [reply lib/compat/wordpress-6.3/template-parts-preloading.php:26]: Sounds like you're seeing different things to me. I'll refine the testing criteria and advise further here.

**Comment by getdave** [comment]
Ok folks this is very strange - the original loading problem I was seeing is now not reproduceable. It's possible something changed underneath me in which case great because this no longer seems to be an issue.

I'm going to refocus my attention at the [Nav block specific loading issues](https://github.com/WordPress/gutenberg/issues/49155).
